### PR TITLE
feat: faster `take` for `BitPackedArray` and `SparseArray`

### LINKED ISSUE #1039: get rid of `relative_indices` in `BitPackedArray::take`
It's likely much faster (no iterators, no heap allocation) to simply scan forward on the next `UNPACK_CHUNK_THRESHOLD` indices (i.e., 8 elements) to see if they're all in a single chunk, then greedily keep taking until we encounter one in a different chunk

### PR DESCRIPTION
**Pull request by lwwmanning**
fixes #1039

### CODE REVIEW

**Comment by lwwmanning** [comment]
Generally the same or much faster across the board

Screenshot 2024-10-24 at 16 44 48

**Comment by gatesn** [comment]
Is the screenshot meant to have 1.0's and question marks?

Or do we have some float rounding errors?

**Comment by lwwmanning** [comment]
> Is the screenshot meant to have 1.0's and question marks?
> 
> Or do we have some float rounding errors?

the 1.0's are normal for `critcmp`. not entirely sure what's up with the `?`s, but vs develop, this is like a 10-70% speed-up, basically across the board

**Comment by lwwmanning** [comment]
Per #1041 , I'm hoping after this that we can maybe revert #1032

File: encodings/fastlanes/src/bitpacking/compute/take.rs [hunk 1]
```diff
@@ -35,111 +35,177 @@ impl TakeFn for BitPackedArray {
...
 
     let packed = array.packed_slice::<T>();
-
     let patches = array.patches().map(SparseArray::try_from).transpose()?;
 
-    // if we have a small number of relatively large batches, we gain by slicing and then patching inside the loop
-    // if we have a large number of relatively small batches, the overhead isn't worth it, and we're better off with a bulk patch
-    // roughly, if we have an average of less than 64 elements per batch, we prefer bulk patching
-    let prefer_bulk_patch = relative_indices.len() * BULK_PATCH_THRESHOLD > indices.len();
+    // Group indices into 1024-element chunks and relativise them to the beginning of each chunk
```
Comment (danking) [inline encodings/fastlanes/src/bitpacking/compute/take.rs:62]: Comment shouldn't say "relativize them" any more as it does not.

**Review by danking — APPROVED** [review]

File: encodings/fastlanes/src/bitpacking/compute/take.rs [hunk 2]
```diff
@@ -35,111 +35,177 @@ impl TakeFn for BitPackedArray {
...
+                PrimitiveArray::from_vec(take_primitive::<$T, $I>(self, &indices)?, taken_validity)
+            })
         });
         Ok(taken.reinterpret_cast(ptype).into_array())
     }
 }
 
-fn take_primitive<T: NativePType + BitPacking>(
+// array_chunks must use while let so that we can get the remainder
+#[allow(clippy::while_let_on_iterator)]
```
Comment (robert3005) [inline encodings/fastlanes/src/bitpacking/compute/take.rs:47]: I think this is correct, you can still use the for loop and then access iterator afterwards

File: vortex-array/src/array/sparse/mod.rs [hunk 3]
```diff
@@ -152,10 +152,22 @@ impl SparseArray {
...
-
             min_index - self.indices_offset()
         })
     }
+
+    /// Return the maximum index if indices are present.
+    ///
+    /// If this sparse array has no indices (i.e. all elements are equal to fill_value)
+    /// then it returns None.
+    pub fn max_index(&self) -> Option<usize> {
```
Comment (robert3005) [inline vortex-array/src/array/sparse/mod.rs:163]: I know I added min_index but we shouldn't pile more here. The methods you want here is first_non_null and last_non_null and that can be implemented for any array

File: vortex-array/src/tree.rs [hunk 4]
```diff
@@ -22,7 +22,7 @@ impl<'a> TreeDisplayWrapper<'a> {
     }
 }
 
-impl<'a> fmt::Display for TreeDisplayWrapper<'a> {
+impl fmt::Display for TreeDisplayWrapper<'_> {
```
Comment (robert3005) [inline vortex-array/src/tree.rs:25]: I think if you're going to run nightly clippy we should do it in general in ci?

**Pull request merged**